entrypoint.sh

```diff
@@ -9,16 +9,38 @@ IFS=';' read -ra paths <<< "$1"
 for i in "${paths[@]}"; do
   echo "processing unit tests in: $i"
   
-  cp /CMakeLists.txt "$i"
+  # Check if CMakeLists.txt already exists in the unit test directory
+  if [ ! -e "$i/CMakeLists.txt" ]; then
+    cp /CMakeLists.txt "$i"
+  fi
+
   pushd "$i" || exit 1
   cmake CMakeLists.txt
   make || exit 2
 
-  # execute the unit tests
+  # Find all executables ending in _tests recursively in the specified root directory
   popd || exit 3
-  valgrind --leak-check=full --error-exitcode=4 "./$i/unit_tests" || exit 4
+  mapfile -t test_executables < <(find "$i" -type f -executable -name "*_tests")
+
+  # Check if any test executables were found
+  if [ ${#test_executables[@]} -eq 0 ]; then
+    echo "no test executables found"
+    exit 4
+  fi
+
+  # execute the unit tests
+  for executable in "${test_executables[@]}"; do
+    valgrind --leak-check=full --error-exitcode=4 "$executable" || exit 4
+  done
+
+  # determine source dir
+  if [ -z "$2" ]; then
+    source_dir="."
+  else
+    source_dir=$2
+  fi
 
   # generate a coverage report
   uuid=$(uuidgen)
-  gcovr --branches --xml-pretty -r . >> "$uuid-report.xml"
+  gcovr . --root "$source_dir"  --branches --xml-pretty >> "$uuid-report.xml"
```

Comment:
gcovr is being given 2 paths, ```.``` and ```"$source_dir```, do we know what gcovr will do with the 2 paths?

note that previously ```.``` was being passed as the root path

Reply:
In this instance, `.` is what gcovr refers to as the *search_paths*, which is where the coverage files are expected to be. `"$source_dir"` is referring to the location of the source files which we want coverage reports for. As per the [gcovr CLI ref](https://gcovr.com/en/master/manpage.html#cmdoption-gcovr-r):

> -r <root>, --root <root>
> The root directory of your source files. Defaults to ‘.’, the current directory. File names are reported relative to this root. The --root is the default --filter. Config key(s): root.

With our configuration of having source files in a different directory to test files, and then targeting them via CMake library, gcovr wasn't able to see the source files it was supposed to be generating reports for. This meant I was just getting information on the tests themselves as well as some gtest code!

Reply:
That makes sense, thanks.